### PR TITLE
[WIP] replace private key identity by mnemonic

### LINKED ISSUE #1: Use mnemonic instead of private key
at the moment we use a single private key as identity﻿. This degrades the privacy for the traders and in general makes the swap looking bad (ie. Address Reuse)

The `wallet` command should ask for a mnemonic seed, the `wallet balance` and `wallet send` should all look for all geerated addresses.

### PR DESCRIPTION
**Pull request by louisinger**
**This PR aims to replace all the Private key identity stuffs by the Mnemonic identity**

TL;DR
- The `walletAction` initializes the wallet using a mnemonic seed (or generate a random one).
- The `walletBalanceAction` fetches the balances of all the addresses derivated from the mnemonic identity.
- The `walletSendAction` uses all the mnemonic addresses to fetch utxos and use a new change address in `buildTx`.
- The `state` stores now the Identity object instead of a single address.

_Why store the identity instead of addresses ?_
We need the identity object to avoid restoration from the explorer each time we need to recreate the Identity object (when we sign the tx for example).

_Why it is a problem ?_
The Identity object contains sensitive data and is not encrypted! Everybody can use it to sign the transactions even if the keystore type is encrypted.

_What are the solutions ?_
**first solution**: encrypt all the state instead of just the private key.
**second solution**: store a set of addresses instead of the identity itself and overwrite the IdentityRestorer to restore addresses from the state each time we recreate the Identity.

@tiero I need your opinion about that

it closes #1

### CODE REVIEW

**Comment by tiero** [comment]
As per discussed offline, I think we should encrypt the whole Identity instance and whenever we need to interact with the wallet (either reading the addresses or signing) we will ask the password to decrypt.

**Comment by tiero** [comment]
In the end I think a custom restorer would be the cleanest solution, for initializing from a set of cached addresses @louisinger

File: src/state.ts [hunk 1]
```diff
@@ -59,26 +83,27 @@ export interface StateNetworkInterface {
...
   selected: boolean;
-  identity?: IdentityInterface;
+  addressesWithBlindingKey: AddressInterface[];
   keystore: {
     type: KeyStoreType;
     value: string;
   };
 }
 
 export function getWalletInfo(walletState: StateWalletInterface): string {
```
Comment (tiero) [inline src/state.ts]: I think this belongs more to helpers than here. I would accept a list of addressesWithBlindingKey directly and craft the string as needed.

File: src/state.ts [hunk 2]
```diff
@@ -169,4 +194,27 @@ export default class State implements StateInterface {
 
     return this.state;
   }
+
+  getMnemonicIdentityFromState(password?: string): Mnemonic {
+    let seed = this.state.wallet.keystore.value;
+    if (this.state.wallet.keystore.type == KeyStoreType.Encrypted) {
+      if (password == null)
```
Comment (tiero) [inline src/state.ts]: !password || password.length === 0

File: src/actions/walletAction.ts [hunk 3]
```diff
@@ -1,150 +1,151 @@
...
+export default async function () {
+  try {
+    info('=========*** Wallet ***==========\n');
+
+    const { network, wallet } = state.get();
+
+    if (!network.selected) return error('Select a valid network');
+
+    // if wallet is already configured, log the state.
+    if (wallet.selected) return log(getWalletInfo(wallet));
```
Comment (tiero) [inline src/actions/walletAction.ts]: But we are not deriving a next address this way. How to deposit funds?

File: src/actions/walletAction.ts [hunk 4]
```diff
@@ -17,8 +17,49 @@ import State, {
...
 const state = new State();
 
+const password = new enquirer.Password({
+  type: 'password',
+  name: 'password',
+  message: 'Type your password',
+});
+
+// wallet: generate a new address
+async function walletAddress() {
```
Comment (tiero) [inline src/actions/walletAction.ts]: lets' move this in a specific wallet action file, so it becomes a proper subcommand like `balance` and `send`

File: src/actions/walletAction.ts [hunk 5]
```diff
@@ -149,3 +183,17 @@ export default async function () {
     error(err);
   }
 }
+
+// command function
+export default async function (walletCommand: string) {
```
Comment (tiero) [inline src/actions/walletAction.ts]: this is not needed. Th library commander already takes care of this if we use subcommand feature

**Review by tiero — COMMENTED** [review]
I tried the CLI locally 

```sh
tdex-cli use_mnemonic_instead_of_private_key 🔒  ✅  node dist/index.js wallet init
=========*** Init wallet ***==========

✔ Want to restore from mnemonic seed? · Nope / Yep
✔ A new wallet will be created. How do you want to store your seed? 🔑 · Encrypted
✔ Type your password · ********

Wallet has been created/restored successfully

Be sure to backup your data directory before sending any funds
ReferenceError: walletState is not defined
```

**Review by tiero — COMMENTED** [review]
I'm doing something wrong, but I cannot make a trade, it give me always `Error: You do not have enough in your wallet` 

Also the balance seems off

this stacktrace from an empty wallet, I only call once the faucet (ie 1 btc) and it gives me  a balance of 200000000


Looking into the explorer directly I only have 1 btc

```
tdex-cli use_mnemonic_instead_of_private_key 🔒  ✅  node ./dist/index.js wallet
✔ Want to restore from mnemonic seed? · Nope / Yep
✔ A new wallet will be created. How do you want to store your seed? 🔑 · Plain

Wallet has been created/restored successfully

Be sure to backup your data directory before sending any funds
0 addresses in your wallet: use "wallet address" to generate new one
tdex-cli use_mnemonic_instead_of_private_key 🔒  ✅  node ./dist/index.js wallet
Wallet is initialized.
0 addresses in your wallet: use "wallet address" to generate new one
tdex-cli use_mnemonic_instead_of_private_key 🔒  ✅  node ./dist/index.js wallet address
new address: el1qqt4j5eldacz0g6uynlv7cu7gkdazfex803drngh4fy80qp7qsqwpq9ctz95lx8wwrg48f7tczy8thn7nepfewftkragwl2jhy
tdex-cli use_mnemonic_instead_of_private_key 🔒  ✅  nigiri faucet --liquid el1qqt4j5eldacz0g6uynlv7cu7gkdazfex803drngh4fy80qp7qsqwpq9ctz95lx8wwrg48f7tczy8thn7nepfewftkragwl2jhy
txId: 387256f760356c348f9922515be361cebf0aef308b89ec0ff19da305e6397e3e
tdex-cli use_mnemonic_instead_of_private_key 🔒  ✅  node ./dist/index.js wallet balance
=========*** Wallet ***==========

*** LBTC ***
Balance 200000000 satoshis
Asset hash: 5ac9f65c0efcc4775e0baec4ec03abdde22473cd3cf33c0419ca290e0751b225

```

File: src/helpers.ts [hunk 6]
```diff
@@ -17,6 +17,24 @@ export const TAXI_API_URL = {
   regtest: 'https://liquid-taxi.herokuapp.com',
 };
 
+export function getWalletInfo(addresses: AddressInterface[]): string {
+  if (addresses.length === 0)
```
Comment (tiero) [inline src/helpers.ts]:
```suggestion
  if (!addresses || addresses.length === 0)
```

**Review by tiero — CHANGES_REQUESTED** [review]
After each trade operation we need to overwrite the internal state of the CLI with the updated list of generated addresses. trade.buy and trade.sell they both derive new addresses to be used in the swap

**Review by tiero — APPROVED** [review]
